Loop/Managers/LoopDataManager.swift

```diff
@@ -926,112 +926,46 @@ extension LoopDataManager {
-                        startingAt: retrospectiveStart,
-                        endingAt: nil,
-                        effectVelocities: settings.dynamicCarbAbsorptionEnabled ? insulinCounteractionEffects : nil
-                    )
-
-                    effects.append(potentialCarbEffect)
-
-                    retrospectiveGlucoseEffect = computeRetrospectiveGlucoseEffect(startingAt: glucose, carbEffects: potentialCarbEffect)
-                }
-            } else if let carbEffect = carbEffectOverride ?? self.carbEffect {
-                effects.append(carbEffect)
-            }
-        }
-
-        if inputs.contains(.insulin) {
-            let computationInsulinEffect: [GlucoseEffect]?
-            if insulinEffectOverride != nil {
-                computationInsulinEffect = insulinEffectOverride
-            } else {
-                computationInsulinEffect = includingPendingInsulin ? self.insulinEffectIncludingPendingInsulin : self.insulinEffect
-            }
-
-            if let insulinEffect = computationInsulinEffect {
-                effects.append(insulinEffect)
-            }
-
-            if let potentialBolus = potentialBolus {
-                guard let sensitivity = insulinSensitivityScheduleApplyingOverrideHistory else {
-                    throw LoopError.configurationError(.generalSettings)
-                }
-
-                let earliestEffectDate = Date(timeInterval: .hours(-24), since: now())
-                let nextEffectDate = insulinCounteractionEffects.last?.endDate ?? earliestEffectDate
-                let bolusEffect = [potentialBolus]
-                    .glucoseEffects(insulinModel: model, insulinSensitivity: sensitivity)
-                    .filterDateRange(nextEffectDate, nil)
-                effects.append(bolusEffect)
-            }
-        }
-
-        if inputs.contains(.momentum), let momentumEffect = self.glucoseMomentumEffect {
-            momentum = momentumEffect
-        }
-
-        if inputs.contains(.retrospection) {
-            effects.append(retrospectiveGlucoseEffect)
-        }
-
-        var prediction = LoopMath.predictGlucose(startingAt: glucose, momentum: momentum, effects: effects)
-
-        // Dosing requires prediction entries at least as long as the insulin model duration.
-        // If our prediction is shorter than that, then extend it here.
-        let finalDate = glucose.startDate.addingTimeInterval(model.effectDuration)
-        if let last = prediction.last, last.startDate < finalDate {
-            prediction.append(PredictedGlucoseValue(startDate: finalDate, quantity: last.quantity))
-        }
-
-        return prediction
+        let pumpStatusDate = doseStore.lastAddedPumpData
+        let effectInterval = carbStore.delta
+        let defaultAbsorptionTime = defaultAbsorptionTime ?? carbStore.defaultAbsorptionTimes.medium // ??? Is this right?
+        
+        return try ForecastGenerator.predictGlucose(startingFrom: now(),
```

Comment:
Yeah, this is pretty crazy. The function arguments here can probably be categorized into three or four categories:
* Therapy settings
* Model inputs (carbs, glucose, and insulin)
* Input effects (these can generally be computed from the inputs)
* Model parameters (adaptiveRateStandbyIntervalFraction, inputDataRecencyInterval, carb delay, etc)

I could see a LoopModel struct with all the model parameters set during construction, with a method `predict()` that took therapy settings and inputs (no effects), and would compute effects and output predicted glucose timeline.

Then that timeline would be passed to the dosing algo in DoseMath.

It seems non-trivial to come up with a pure-refactoring path that goes from where we're at to what I describe above, but I think this functional step might still be a good first step.

Reply:
It does illustrate what kind of crazy test matrix this function calls for.  What I like is that while `ForecastGenerator` is in `Loop`, it only has `LoopKit` as a dependency -- and no dependencies on settings, or "Stores".  

Unless some of these parameters are strictly constants, I would still prefer the functional, stateless approach.  My preference would be to use your grouping to create `structs` and pass them in to better illustrate the behavior of the `predictGlucose` function.

So...how should I proceed, @ps2 ?

Reply:
Yes, I agree on keeping it functional. The "Model parameters" grouping I describe above is static for normal Loop usage; they never change. You could call them meta-parameters maybe?

Therapy Settings is an easy grouping, I think. And you could try grouping model inputs and carbs, though eventually one should produce the other. The last group, like I said, is basically static. 

Yes, keeping the stores out of this is great!